packages/firestore/src/remote/persistent_stream.ts

```diff
@@ -435,49 +447,45 @@ export abstract class PersistentStream<
       });
     };
 
-    // Only start stream if listener has not changed
-    if (this.listener !== null) {
-      const currentStream = this.startRpc(token);
-      this.stream = currentStream;
-      this.stream.onOpen(() => {
-        dispatchIfStillActive(currentStream, () => {
-          assert(
-            this.state === PersistentStreamState.Auth,
-            'Expected stream to be in state auth, but was ' + this.state
-          );
-          this.state = PersistentStreamState.Open;
-          return this.listener!.onOpen();
-        });
+    const currentStream = this.startRpc(token);
+    this.stream = currentStream;
+    this.stream.onOpen(() => {
+      dispatchIfStillActive(currentStream, () => {
+        assert(
+          this.state === PersistentStreamState.Starting,
+          'Expected stream to be in state Starting, but was ' + this.state
+        );
+        this.state = PersistentStreamState.Open;
+        return this.listener!.onOpen();
       });
-      this.stream.onClose((error: FirestoreError) => {
-        dispatchIfStillActive(currentStream, () => {
-          return this.handleStreamClose(error);
-        });
+    });
+    this.stream.onClose((error: FirestoreError) => {
+      dispatchIfStillActive(currentStream, () => {
+        return this.handleStreamClose(error);
       });
-      this.stream.onMessage((msg: ReceiveType) => {
-        dispatchIfStillActive(currentStream, () => {
-          return this.onMessage(msg);
-        });
+    });
+    this.stream.onMessage((msg: ReceiveType) => {
+      dispatchIfStillActive(currentStream, () => {
+        return this.onMessage(msg);
       });
-    }
+    });
   }
 
-  private performBackoff(listener: ListenerType): void {
+  private performBackoff(): void {
     assert(
       this.state === PersistentStreamState.Error,
-      'Should only perform backoff in an error case'
+      'Should only perform backoff when in Error state'
     );
     this.state = PersistentStreamState.Backoff;
 
     this.backoff.backoffAndRun(async () => {
-      if (this.state === PersistentStreamState.Stopped) {
-        // We should have canceled the backoff timer when the stream was
-        // closed, but just in case we make this a no-op.
-        return;
-      }
+      assert(
```

Comment:
Interesting. I see what you mean re inherent safety.

On the other hand, I'd like to get us out of the business of having to make that kind of argument when reasoning about this code. I would prefer to just handle all the callbacks in a uniform manner so that we don't have to think about it.